PHX-219 Subsection Gating

douglashall opened the pull request:
This PR implements the Subsection Gating feature described in the following Confluence document:

https://openedx.atlassian.net/wiki/display/SOL/Subsection+Gating

LMS Sandbox URL:
https://subsection-gating.sandbox.edx.org/courses/course-v1:ASUx+MAT117+2016_T2/courseware/d6cd7bf7d98e48aba1cd3f410ec88ca8/

Studio Sandbox URL:
https://studio-subsection-gating.sandbox.edx.org/course/course-v1:ASUx+MAT117+2016_T2

PR Dependencies:
https://github.com/edx/edx-milestones/pull/18

chrisndodge commented:
wow, congrats. Looks interesting. I'll look deeper tomorrow.

robrap commented:
Cool. I haven't had a chance to review code yet, just the docs. They mention avoiding circularity, but not broken requirements if one side of a dependency is deleted. Is that scenario handled here or via milestone api?

douglashall commented:
@robrap The circular dependency check has not yet been implemented, but I think that would live within the milestones service. As far as deletion goes, I think the client of the milestone service would be responsible for handling that condition. The gating feature in this PR handles the removal of a prerequisite subsection (the 'fulfills' side of the milestone relationship) by calling milestones.api.remove_milestone which in turn soft-deletes both sides (fulfills/requires) of the associated relationships. I could envision other use cases that would want to do something different to the underlying milestone data.

robrap commented:
@douglashall Thanks for the response about deletion.

douglashall commented:
jenkins run python

douglashall commented:
jenkins run lettuce

douglashall commented:
jenkins run lettuce

douglashall commented:
@talbs I have updated this branch with the changes to the studio flow that we discussed. Let me know if you need any pointers to the front-end code. Also, I updated the sandbox with these changes:

https://studio-subsection-gating.sandbox.edx.org/course/course-v1:ASUx+MAT117+2016_T2

afzaledx commented on cms/templates/js/course-outline-modal.underscore:
```diff
@@ -1,14 +1,6 @@
-        </div>
-    <% } %>
+    <div class="message modal-introduction">
+        <p><%- introductionMessage %></p>
```
We were asked to remove the Introduction Message which starts with "Change the settings for ..." (https://github.com/edx/edx-platform/pull/10569#issuecomment-163691095) when showing multiple tabs. Removing the above checks always shows the introduction message. I think we should probably still show the introduction message when tab-count==1, and should not show it when >= 2 tabs

douglashall replied:
Yes, sorry for the back and forth on this, but I think we want the intro message in the second redesign. @talbs Can you confirm this?

talbs replied:
Sorry for the back and forth, @douglashall and @afzaledx. Looking at the updated sandbox, I think we can remove the "Change the settings for ..." text. Its not needed given the rest of the UI's context. Thanks for the attention to minor details!

afzaledx commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -751,6 +762,37 @@ def _get_module_info(xblock, rewrite_static_links=True, include_ancestor_info=Fa
+    info = {}
+    if xblock.category == 'sequential':
+        # Gating info
+        if course.enable_subsection_gating:
```
line 498, the same check is written in a different way:  if xblock.location.category == 'sequential' and course.enable_subsection_gating

xblock.location.category vs xblock.category, and 2-lines vs single line. Perhaps you can choose one over the other for both places?

douglashall replied:
Yes. I will make those consistent. Thanks.

afzaledx commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -488,6 +475,52 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+            var showPrereqInput = this.$('#prereq option:selected').val().length > 0;
+            this.$('#prereq_min_score_input').toggle(showPrereqInput);
+        },
+        validate: function () {
```
nit: A more specific name for the validator method since we're validating just the required min score here. e.g. validateScore

douglashall replied:
Sounds good to me.

afzaledx commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -488,6 +475,52 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                minScore &&
+                (
+                    typeof(minScoreInt) === 'undefined' ||
+                    String(minScoreInt) !== minScore ||
```
Maybe we should trim the spaces as well before the check.
parseInt would accept "40 "(a space after 40) as valid, but since "40" and "40 " are not equal, this would make the number invalid. Especially use when someone's copy-pasting and copies an additional space character, and has no clue where the error lies. Moreover, since I am a regex fanboy, I'd do this validation with a regex :-).

douglashall replied:
Yes, thank you.

afzaledx commented:
:+1: other than minor suggestions.

douglashall commented:
@afzaledx I have updated this PR based on your feedback. Thanks.

talbs commented:
@douglashall, thanks for the UI tweaks. This looks good overall. A few small nits/changes:

![image](https://cloud.githubusercontent.com/assets/163763/12051858/ca9c087e-aed6-11e5-81dc-013d7eca4bcf.png)

The alignment of the Advanced setting fields seems off with the titles (see screengrab above)

- - -

I've updated some of the copy/labels on the "Access" tab as well as on the Outline gating note and the deletion modal. I'll make notes within the lines of code as well. Here are some quick screengrabs for reference:

![image](https://cloud.githubusercontent.com/assets/163763/12051891/412ee16e-aed7-11e5-9710-009c8af04152.png)

![image](https://cloud.githubusercontent.com/assets/163763/12051932/cb8cb188-aed7-11e5-8de5-e8dacfd1258e.png)

![image](https://cloud.githubusercontent.com/assets/163763/12052018/b703776e-aed8-11e5-81fd-24c21916db08.png)

- - -

Once you address those bits, this looks good to me and you have my :+1:

talbs commented on cms/templates/js/course-outline.underscore:
```diff
@@ -2,9 +2,21 @@
+        }
+    });
+    statusType = 'staff-only';
+    statusMessage = gettext('Gated on ' + prereqDisplayName);
```
NIT: change copy to "Prerequisite: "

talbs commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,48 @@
+            </ul>
+        </div>
+    <% } %>
+    <h3 class="modal-section-title"><%= gettext('Set as a Prerequisite') %></h3>
```
NIT: Change copy to "Use this as a Prerequisite"

talbs commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,48 @@
+                <label for="is_prereq" class="label">
+                    <i class="icon fa fa-check-square-o input-checkbox-checked"></i>
+                    <i class="icon fa fa-square-o input-checkbox-unchecked"></i>
+                    <%- gettext('Yes, set as a prerequisite') %>
```
NIT: Change copy to "Yes, set this as a prerequisite other content can limit access with"

douglashall replied:
@talbs How about "Yes, set this as a prerequisite which can be used to limit access to other content"?

douglashall replied:
@pdesjardins Can you weigh in on the wording here?

douglashall commented:
Thanks @talbs! I have incorporated your feedback and updated the sandbox.

https://studio-subsection-gating.sandbox.edx.org/course/course-v1:ASUx+MAT117+2016_T2

cahrens commented:
@douglashall I will be reviewing this. The new tabs in the course settings look really nice! If you haven't already, make sure it gets a quick accessibility review. We would also like to see a11y bok choy tests with new features. I promise they are very easy to write, and they have been very useful so far. 

See https://openedx.atlassian.net/wiki/display/TE/Automated+Accessibility+Tests#AutomatedAccessibilityTests-WritinganAccessibilityTest and/or search in the acceptance test directory for "check_for_accessibility_errors".

cahrens commented:
Currently the following limitation exists-- if a student fulfills a prerequisite section (for instance, gets a high enough score on an exam), the gated content does not appear until the student refreshes the browser. This should certainly be documented for course staff as a limitation, as it seriously impacts the student experience for use cases where the student can complete a milestone without any staff/LTI interaction.

cahrens commented on lms/djangoapps/courseware/views.py:
```diff
@@ -402,6 +397,14 @@ def _index_bulk_op(request, course_key, chapter, section, position):
             return redirect(reverse('courseware', args=[unicode(course.id)]))
+
+    # Gated Content Check
+    gated_content = gating_api.get_gated_content(course.id, user)
```
This will need to be load-tested to make sure that it doesn't adversely impact LMS page loading time.

https://openedx.atlassian.net/wiki/display/EdxOps/How+to+Run+Performance+Tests

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -497,9 +494,22 @@ def _save_xblock(user, xblock, data=None, children_strings=None, metadata=None,
-        # value of the staff lock will also update the published version, but only at the unit level.
+        # Save gating info
+        if xblock.category == 'sequential' and course.enable_subsection_gating:
+            course_key = xblock.runtime.course_id.for_branch(None)
```
Why can't xblock.location.course_key be used as above?

douglashall replied:
It can be. Thanks.

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -163,6 +157,9 @@ def xblock_handler(request, usage_key_string):
                 nullout=request.json.get('nullout'),
                 grader_type=request.json.get('graderType'),
+                is_prereq=request.json.get('isPrereq'),
+                prereq_usage_key=request.json.get('prereqUsageKey'),
```
Update documentation above to include these new fields.

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -657,6 +667,7 @@ def _delete_item(usage_key, user):
             store.update_item(course, user.id)
 
         store.delete_item(usage_key, user.id)
+        SignalHandler.item_deleted.send(sender=None, usage_key=usage_key, user_id=user.id)
```
Does store.delete_item not emit a signal (or should it)? I'm not familiar with this SignalHandler, but it seems odd that it would be necessary to call both store.delete_item and the item_deleted signal handler.

douglashall replied:
Yes, I need to push signal emission down to store.delete_item. Thanks.

cahrens commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -751,6 +762,35 @@ def _get_module_info(xblock, rewrite_static_links=True, include_ancestor_info=Fa
+        info["prereq"] = prereq
+        info["prereq_min_score"] = prereq_min_score
+        if prereq:
+            info["visibility_state"] = VisibilityState.staff_only
```
A new visibility state should be created. Returning "staff_only" is very confusing (even if we wish to render them in similar ways).

cahrens commented:
Take a look at the missing unit test coverage to see if the new gated content code can be covered. https://build.testeng.edx.org/job/edx-platform-js-pr/11182/Diff_Coverage_Report/ and https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/11308/Diff_Coverage_Report/

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -624,6 +634,15 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
                 };
 
+                selectPrerequisite = function() {
+                    this.$("#is_prereq").prop('checked', true).trigger('change');
```
Is it not possible to actually click the button?

douglashall replied:
It is not because we are using some css magic to display a fancy icon based checkbox.

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -988,6 +1020,133 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
+                    createCourseOutlinePage(this, mockCourseJSON, false);
+                    outlinePage.$('.outline-subsection .configure-button').click();
+                    selectPrerequisite();
+                    $('.wrapper-modal-window .action-save').click();
```
I don't see that anything is asserted here after clicking the prerequisite. What is being tested?

cahrens commented:
Even with a validation error, I am able to save. I then get the generic sever error bar at the bottom (in this picture, I have already saved once with validation error, which is why the bar and the dialog show at the same time).

![image](https://cloud.githubusercontent.com/assets/484484/12094622/79d63e50-b2d7-11e5-95e7-c2d5e4d6a6e7.png)

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -988,6 +1020,133 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
+                    selectLastPrerequisiteSubsection('abc');
+                    expect($('#prereq_min_score_error').css('display')).not.toBe('none');
+                    selectLastPrerequisiteSubsection('5.5');
+                    expect($('#prereq_min_score_error').css('display')).not.toBe('none');
```
I was surprised that validation errors do not prevent saving.

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -488,6 +479,56 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+        templateName: 'access-editor',
+        className: 'edit-settings-access',
+        events : {
+            'change #prereq': 'handlePrereqSelect',
```
As a general note, we prefer classes to ids (because of ids can cause non-uniqueness issue). Since this is a modal, it is extremely unlikely 2 will ever be shown at the same time, but in the future please use classes instead of ids.

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -687,20 +728,32 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
-                editors = [ReleaseDateEditor, GradingEditor, DueDateEditor];
+                tabs.push({
+                    name: 'basic',
+                    displayName: 'Basic',
```
i18n

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -687,20 +728,32 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                if (options.enable_proctored_exams || options.enable_timed_exams) {
+                    tabs.push({
+                        name: 'advanced',
+                        displayName: 'Advanced',
```
i18n

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -687,20 +728,32 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                if (typeof(xblockInfo.get('is_prereq')) !== 'undefined') {
+                    tabs.push({
+                        name: 'access',
+                        displayName: 'Access',
```
i18n

cahrens commented on cms/static/js/views/utils/xblock_utils.js:
```diff
@@ -73,13 +73,23 @@ define(["jquery", "underscore", "gettext", "common/js/components/utils/view_util
+                additionalBodyText = '';
+            xblockType = xblockType || gettext('component');
+            if (xblockInfo.get('is_prereq')) {
+                additionalHeaderText = gettext(' (and prerequisite)');
```
You can't concatenate together text like this-- it simply won't work for translation. You need to use interpolation and placeholders.

cahrens commented on cms/static/js/views/utils/xblock_utils.js:
```diff
@@ -73,13 +73,23 @@ define(["jquery", "underscore", "gettext", "common/js/components/utils/view_util
+            if (xblockInfo.get('is_prereq')) {
+                additionalHeaderText = gettext(' (and prerequisite)');
+                additionalBodyText = gettext(
+                    ' Any content that has listed this content as a prerequisite ' +
```
Don't break up strings in JavaScript that will be translated into multiple lines-- I'm pretty sure the i18n scraping fails on this.

Note that you should test your UI with mock translation done to verify that strings are getting scraped properly. See "Coverage testing" here: http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html.

cahrens commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+<form>
+    <% if (xblockInfo.get('prereqs').length > 0) { %>
+        <h3 class="modal-section-title"><%= gettext('Limit Access') %></h3>
```
Use %- instead of %=

cahrens commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+                        <%- gettext('Prerequisite:') %>
+                    </label>
+                    <select id="prereq" class="input">
+                        <option value="">No prerequisite</option>
```
i18n

cahrens commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+                <p class="field-message">
+                    <%- gettext('Select a prerequisite subsection and enter a minimum score percentage to limit access to this subsection.') %>
+                </p>
+                <li class="field field-select">
```
Why are these being rendered as list items?

douglashall replied:
I'm following the style used in the existing editor templates.

cahrens commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+                    <label for="prereq_min_score" class="label">
+                        <%- gettext('Minimum Score:') %>
+                    </label>
+                    <input type="text" id="prereq_min_score" name="prereq_min_score" class="input input-text" size="3" aria-describedby="Minimum score must be an integer between 0 and 100." placeholder="0 - 100" />
```
aria-describedby should be an ID name, not free-form text.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute

I suggest you write your bok choy accessibility test before fixing this bug, as it should fail (with an error that there is no element with this ID).

cahrens commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+                    </select>
+                </li>
+                <li id="prereq_min_score_input" class="field field-input input-cosmetic">
+                    <label for="prereq_min_score" class="label">
```
You can nest the input within the label to avoid the need for linking by ID. This is covered in the accessibility edX course.

For example,
````
<label>
   <input type="text" name="lastname" />
   Last Name
</label>
````

cahrens commented on cms/templates/js/access-editor.underscore:
```diff
@@ -0,0 +1,45 @@
+    <% } %>
+    <h3 class="modal-section-title"><%= gettext('Use this as a Prerequisite') %></h3>
+    <div class="modal-section-content gating-is-prereq">
+        <ul class="list-fields list-input">
```
Why is a list being used here?

cahrens commented on cms/templates/js/course-outline-modal.underscore:
```diff
@@ -1,14 +1,6 @@
-            <p><%- introductionMessage %></p>
-        </div>
-    <% } %>
+    <div class="message modal-introduction">
```
Is this message being removed entirely per @talbs comment below?

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -2,9 +2,21 @@
+            return false;
+        }
+    });
+    statusType = 'staff-only';
```
Again, it is odd to be using "staff-only" here. Let's create a different statusType, even if it is rendered the same way.

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -2,9 +2,21 @@
+        }
+    });
+    statusType = 'staff-only';
+    statusMessage = gettext('Prerequisite: ' + prereqDisplayName);
```
i18n-- need to use interpolation

cahrens commented on cms/templates/js/settings-modal-tabs.underscore:
```diff
@@ -0,0 +1,10 @@
+<ul class="settings-tabs-header">
+    <% _.each(tabs, function(tab) { %>
+    <li class="settings-tab-buttons">
+        <button class="settings-tab-button" data-tab="<%- tab.name %>" href="#"><%- gettext(tab.displayName) %></button>
```
The gettext needs to be done where the displayName is stored so that the string is scraped.

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -1368,6 +1379,18 @@ def can_toggle_course_pacing(self):
+        Returns:
+            list: Returns a list of dicts containing the gating milestone data
+        """
+        if not self._gating_prerequisites:
```
Can you explain the caching you are doing here? Would gating_api.get_prerequisites(self.id) ever need to be updated within the same session? Why is this not done in init?

douglashall replied:
This keeps us from having to make a database call to retrieve gating milestone data for every content node in the course tree when rendering the studio course outline. We only need one database call to find all gating prerequisites for the whole course for a given request. I suppose this could be moved to init where we could check the `enable_subsection_gating` flag on the course, but I didn't want to make the extra database call unless we needed it.

cahrens commented on common/test/acceptance/pages/studio/overview.py:
```diff
@@ -635,6 +636,75 @@ def proctoring_items_are_displayed(self):
+        if not self.q(css="#is_prereq")[0].is_selected():
+            self.q(css='label[for="is_prereq"]').click()
+        self.q(css=".action-save").first.click()
+        self.wait_for_ajax()
```
We have found that wait_for_ajax is problematic-- relying on it often results in a flaky test. If there is a specific thing on the page that you can wait for to signify that the operation is complete, that is much more reliable.

I recommend that you run your new tests many tests with the flaky decorator (min_pass of 20 or 30) before submitting them to ensure they are robust. Remember to remove the flaky decorator before submission though!

cahrens commented on common/test/acceptance/pages/studio/overview.py:
```diff
@@ -635,6 +636,75 @@ def proctoring_items_are_displayed(self):
+        """
+
+        # The Prerequisite checkbox is visible
+        if not self.q(css="#is_prereq").visible:
```
Simplify to self.q(css="#is_prereq").visible.

cahrens commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -589,6 +589,9 @@ def _progress_summary(student, request, course, field_data_cache=None, scores_cl
         max_scores_cache.fetch_from_remote(field_data_cache.scorable_locations)
 
+    # Check for gated content
+    gated_content = gating_api.get_gated_content(course.id, student)
```
Grading load tests should be run.

cahrens commented on lms/djangoapps/courseware/views.py:
```diff
@@ -402,6 +397,14 @@ def _index_bulk_op(request, course_key, chapter, section, position):
+    if section and gated_content:
+        for usage_key in gated_content:
+            if section in usage_key:
+                raise Http404
```
I don't think this has any unit test coverage.

cahrens commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -182,6 +182,10 @@ def toc_for_course(user, request, course, active_chapter, active_section, field_
                           section.url_name == active_section)
 
+                # Skip the current section if it is gated
+                if gated_content and unicode(section.location) in gated_content:
```
I don't think this has any unit test coverage.

cahrens commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -156,9 +152,13 @@ def toc_for_course(user, request, course, active_chapter, active_section, field_
         required_content = milestones_helpers.get_required_content(course, user)
 
+        # Check for gated content
+        gated_content = gating_api.get_gated_content(course.id, user)
```
Load testing needs to be considered here.

cahrens commented on lms/djangoapps/gating/api.py:
```diff
@@ -0,0 +1,91 @@
+            sequential.location.for_branch(None),
+            'fulfills'
+        )
+        print prereq_milestone
```
print call

cahrens commented on lms/djangoapps/gating/api.py:
```diff
@@ -0,0 +1,91 @@
+                    gated_content_milestones[milestone_id] = gated_content
+                gated_content.append(milestone)
+
+            gated_content = gated_content_milestones.get(prereq_milestone['id'])
```
This code confuses me because gated_content is appended to at line 68, but then is set to something else here?

cahrens commented on lms/djangoapps/gating/api.py:
```diff
@@ -0,0 +1,91 @@
+
+            gated_content = gated_content_milestones.get(prereq_milestone['id'])
+            if gated_content:
+                from courseware.grades import get_module_score
```
Why the late import? Is there a circular reference being worked around? If so, please see if there is any way to get rid of the circular reference.

cahrens commented on lms/djangoapps/gating/api.py:
```diff
@@ -0,0 +1,91 @@
+                user = User.objects.get(id=user_id)
+                score = get_module_score(user, course, sequential) * 100
+                for milestone in gated_content:
+                    # Default minimum grade to 0
```
Comment does match line of code below it.

cahrens commented on lms/djangoapps/gating/api.py:
```diff
@@ -0,0 +1,91 @@
+                from courseware.grades import get_module_score
+                user = User.objects.get(id=user_id)
+                score = get_module_score(user, course, sequential) * 100
+                for milestone in gated_content:
```
I am confused how there can be multiple milestones (each with a possible min_score), but only a single computed score that is used to compare against all of them. Can you explain?

douglashall replied:
`milestone` actually represents a CourseContentMilestone (relationship between a milestone and a piece of course content) here. The min score is assigned to the 'requires' side of this relationship. A given piece of content could have multiple milestones which it fulfills, so here we are finding all the milestones which the given content fulfills, then finding all the content that requires those milestones, and comparing the score achieved on the given content with the min score configured for each of the 'requires' relationships associated with a given milestone.

cahrens commented on lms/djangoapps/gating/api.py:
```diff
@@ -0,0 +1,91 @@
+                        try:
+                            min_score = int(requirements.get('min_score'))
+                        except (ValueError, TypeError):
+                            log.warning(
```
Have you considered if there should be any eventing around gated content? Perhaps eventing of milestones already exists and is sufficient, but I would recommend checking in with analytics.

cahrens commented on openedx/core/lib/gating/api.py:
```diff
@@ -0,0 +1,306 @@
+    Raises:
+        GatingValidationError: If the minimum score is not valid
+    """
+    message = "{min_score} is not a valid grade percentage"
```
i18n

cahrens commented on openedx/core/lib/gating/api.py:
```diff
@@ -0,0 +1,306 @@
+def set_required_content(course_key, gated_content_key, prereq_content_key, min_score):
+    """
+    Adds a `requires` milestone relationship for the given gated_content_key if a prerequisite
+    prereq_content_key is provided. If prereq_content_key is None, removes the `requires`
```
I don't see that the code removes the requires relationship if prereq_content_key is None (which seems like a bad idea to me anyway, but the doc should be updated).

cahrens commented on openedx/core/lib/gating/api.py:
```diff
@@ -0,0 +1,306 @@
+        gated_content_key (str|UsageKey): The gated content usage key
+
+    Returns:
+        tuple: The prerequisite content usage key and minimum score, (None, None) the the content is not gated
```
typo

cahrens commented:
@douglashall I'm done with my review. I had a lot of comments and ran over our 2-hour allocated time, so I just did a cursory review of the django gating package. It would be good for the first reviewer to also take a look at my comments, as I think that a lot of this could have been caught in the first review.

My largest concerns are around ensuring that LMS page load time does not increase due to the milestone checking, adding bok choy accessibility tests, adding missing unit tests coverage, and fixing the i18n issues.

douglashall commented:
jenkins run js

douglashall commented:
@cahrens https://openedx.atlassian.net/wiki/display/SOL/Subsection+Gating+Load+Testing

douglashall commented:
jenkins run bokchoy

douglashall commented:
jenkins run bokchoy

douglashall commented:
jenkins run bokchoy

douglashall commented:
@cahrens I believe I have a addressed items from your review and completed load testing. Anything else I can do to get the thumb?

cahrens commented:
@douglashall Thanks for all the changes that you made and doing the load tests!

You don't need a thumbs-up from me. The process for TNL second reviews of Solutions PRs is that we provide the feedback and trust that you guys will follow up appropriately.

I do highly recommend bok choy accessibility tests though (see comment above where I provided some information about them). We have found them to be very easy to write, and they have caught a lot of issues, including JS bugs in some cases. All you have to do is navigate to a page where the component being tested is visible, and call "check_for_accessibility_errors" on your existing page object.

douglashall commented:
@cahrens Excellent, thank you.
